### PR TITLE
Mostafa dev

### PR DESCRIPTION
**Pull request by MostafaRahgouy**
@Rahgooy, the build_vocab script updated and it's ready to use for the error analyzation!

### CODE REVIEW

File: data_helpers.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import numpy as np
+import Read_data_and_Write_results as rw
+from nltk.stem import PorterStemmer as stemmer
+import vocabs as v
+
+class Read:
```
Comment (Rahgooy) [inline data_helpers.py]: how about `Reader`?

File: vocabs.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import Read_data_and_Write_results as rw
+from nltk.stem import PorterStemmer as stemmer
+import json
+import codecs
+
+def build_vocab(word_lsts):
+    words = np.unique([stemmer().stem(word.lower()) for word in np.array(word_lsts).ravel()])
+    Id2word = dict(enumerate(words))
+    word2Id = dict(map(reversed, Id2word.items()))
+    return Id2word , word2Id
```
Comment (Rahgooy) [inline vocabs.py:12]: Wow!

File: vocabs.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+def get_dataId2word(word_lsts, words_id):
+    ids = np.array(words_id).ravel()
+    words = [stemmer().stem(word.lower()) for word in np.array(word_lsts).ravel()]
+    return dict(zip(ids, words))
```
Comment (Rahgooy) [inline vocabs.py:17]: This is neat!

File: vocabs.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def report_new_words(vocab_path, text_path, output_path):
+    
+    with codecs.open(vocab_path, "r", encoding = "utf-8") as f:
+        model = f.read()
+    vocab = {item.split("\t")[0]:item.split("\t")[1] for item in model.split("\n") if len(item) != 0}
+
```
Comment (Rahgooy) [inline vocabs.py]: Here we just work with the words, so why not a `set`?

File: word_conditional_model.py [hunk 5]
```diff
@@ -9,59 +9,22 @@
...
-            for j in range(len(words_lsts[i])):
-                d = self.__words.index( stemmer().stem(words_lsts[i][j].lower()))
-                for m in range(9):
-                    if bios_lsts[i][j][m * 2] == "I" or bios_lsts[i][j][m * 2] == "B":
-                        matrix[(i * 9) + m][d] = 1
-        return matrix
-
-    def __calculate_condetional_probs(self, w_i, w_j):
+    def __calculate_condetional_probs(self,x, y, w_i, w_j):
         occurrenece_of_wi_and_wj, bold_wi = 0, 0
```
Comment (Rahgooy) [inline word_conditional_model.py]: remove this line!

File: word_conditional_model.py [hunk 6]
```diff
@@ -9,59 +9,22 @@
...
-        occurrenece_of_wi_and_wj = (self.__words_matrix[:, w_i] * self.__words_matrix[:, w_j]).sum()
-        y = self.__words_matrix[:, w_i] * self.__words_matrix[:, w_j]
-        bold_wi = (y * self.__bios_matrix[:, w_i]).sum()
+        occurrenece_vector = x[:, w_i] * x[:, w_j]
+        occurrenece_of_wi_and_wj = occurrenece_vector.sum()
```
Comment (Rahgooy) [inline word_conditional_model.py]: `wi_wj_count = (x[:, w_i] * x[:, w_j]).sum()` to be shorter :)

Comment (Rahgooy) [inline word_conditional_model.py]:
```python
wi_wj = x[:, w_i] * x[:, w_j]
wi_wj_count = wi_wj.sum()
```
To be shorter ;)

Comment (MostafaRahgouy) [reply word_conditional_model.py]: Did it!

File: data_helpers.py (hunk 1 above)
Comment (MostafaRahgouy) [reply data_helpers.py]: `read` has been changed to `reader`.

File: vocabs.py (hunk 3 above)
Comment (MostafaRahgouy) [reply vocabs.py:17]: Yeah, I think so too!

File: word_conditional_model.py (hunk 5 above)
Comment (MostafaRahgouy) [reply word_conditional_model.py]: Did it!

File: vocabs.py (hunk 4 above)
Comment (MostafaRahgouy) [reply vocabs.py]: It changed!

File: data_helpers.py [hunk 7]
```diff
@@ -3,35 +3,48 @@
...
-        for i in range(len(words_lsts)):
-            for j in range(len(words_lsts[i])):
-                wordId = word2Id[stemmer().stem(words_lsts[i][j].lower())]
-                matrix[i*9:(i+1)*9, wordId] = 1
+    def get_one_hot_matrix(mtx, word2Id, lenght = 9):
+        matrix = build_matrix(len(mtx.shape[0] * lenght), len(word2Id))
+        for i in range(len(mtx)):
+            temp = [Id for Id in mtx[i] if Id != -1]
+            matrix[i*lenght:(i+1)*lenght, temp] = 1
         return matrix
```
Comment (Rahgooy) [inline data_helpers.py]: This should go outside `read`

File: data_helpers.py [hunk 8]
```diff
@@ -3,35 +3,48 @@
...
 
+    def build_matrix(number_of_sentences, number_of_words):
+        word_matrix = np.zeros((number_of_sentences, number_of_words))
+        return word_matrix
```
Comment (Rahgooy) [inline data_helpers.py]: We don't need this function it is just one line

**Pull request merged**